File: backend/services/medication-service/flow2-go-engine/internal/models/snapshot_models.go
package models

import (
	"time"
)

// SnapshotRequest represents a request to create a clinical snapshot
type SnapshotRequest struct {
	PatientID       string `json:"patient_id" binding:"required"`
	RecipeID        string `json:"recipe_id" binding:"required"`
	ProviderID      string `json:"provider_id,omitempty"`
	EncounterID     string `json:"encounter_id,omitempty"`
	TTLHours        int    `json:"ttl_hours" binding:"min=1,max=24"`
	ForceRefresh    bool   `json:"force_refresh"`
	SignatureMethod string `json:"signature_method,omitempty"`
}

// ClinicalSnapshot represents an immutable clinical snapshot
type ClinicalSnapshot struct {
	ID                string                 `json:"id"`
	PatientID         string                 `json:"patient_id"`
	RecipeID          string                 `json:"recipe_id"`
	ContextID         string                 `json:"context_id"`
	Data              map[string]interface{} `json:"data"`
	CompletenessScore float64                `json:"completeness_score"`
	Checksum          string                 `json:"checksum"`
	Signature         string                 `json:"signature"`
	SignatureMethod   string                 `json:"signature_method"`
	Status            string                 `json:"status"`
	CreatedAt         time.Time              `json:"created_at"`
	ExpiresAt         time.Time              `json:"expires_at"`
	AccessedCount     int                    `json:"accessed_count"`
	LastAccessedAt    *time.Time             `json:"last_accessed_at,omitempty"`
	ProviderID        string                 `json:"provider_id,omitempty"`
	EncounterID       string                 `json:"encounter_id,omitempty"`
	AssemblyMetadata  map[string]interface{} `json:"assembly_metadata"`
	EvidenceEnvelope  map[string]interface{} `json:"evidence_envelope"`
}

// IsExpired checks if the snapshot has expired
func (cs *ClinicalSnapshot) IsExpired() bool {
	return time.Now().After(cs.ExpiresAt)
}

// IsValid checks if the snapshot is valid (not expired and active)
func (cs *ClinicalSnapshot) IsValid() bool {
	return cs.Status == "active" && !cs.IsExpired()
}

// SnapshotValidationResult represents the result of snapshot validation
type SnapshotValidationResult struct {
	SnapshotID            string    `json:"snapshot_id"`
	Valid                 bool      `json:"valid"`
	ChecksumValid         bool      `json:"checksum_valid"`
	SignatureValid        bool      `json:"signature_valid"`
	NotExpired           bool      `json:"not_expired"`
	Errors               []string  `json:"errors"`
	Warnings             []string  `json:"warnings"`
	ValidatedAt          time.Time `json:"validated_at"`
	ValidationDurationMs float64   `json:"validation_duration_ms"`
}

// SnapshotBasedFlow2Request represents a snapshot-based Flow2 execution request
type SnapshotBasedFlow2Request struct {
	// Option 1: Use existing snapshot
	SnapshotID string `json:"snapshot_id,omitempty"`

	// Option 2: Create new snapshot
	PatientID         string   `json:"patient_id" binding:"required"`
	RecipeID          string   `json:"recipe_id,omitempty"`
	MedicationCode    string   `json:"medication_code" binding:"required"`
	MedicationName    string   `json:"medication_name,omitempty"`
	Indication        string   `json:"indication,omitempty"`
	PatientConditions []string `json:"patient_conditions,omitempty"`
	Priority          string   `json:"priority,omitempty"`

	// Snapshot creation options (when creating new snapshot)
	ProviderID   string `json:"provider_id,omitempty"`
	EncounterID  string `json:"encounter_id,omitempty"`
	TTLHours     int    `json:"ttl_hours,omitempty"`
	ForceRefresh bool   `json:"force_refresh,omitempty"`

	// Processing options
	ProcessingHints map[string]interface{} `json:"processing_hints,omitempty"`
}

// SnapshotBasedFlow2Response represents the response from snapshot-based execution
type SnapshotBasedFlow2Response struct {
	RequestID      string                        `json:"request_id"`
	PatientID      string                        `json:"patient_id"`
	SnapshotInfo   *SnapshotInfo                 `json:"snapshot_info"`
	IntentManifest *IntentManifestResponse       `json:"intent_manifest"`
	MedicationProposal *MedicationProposal       `json:"medication_proposal"`
	EvidenceEnvelope *SnapshotEvidenceEnvelope   `json:"evidence_envelope"`
	OverallStatus    string                      `json:"overall_status"`
	PerformanceMetrics *SnapshotPerformanceMetrics `json:"performance_metrics"`
	Timestamp        time.Time                   `json:"timestamp"`
}

// SnapshotInfo contains information about the snapshot used
type SnapshotInfo struct {
	SnapshotID        string    `json:"snapshot_id"`
	RecipeID          string    `json:"recipe_id"`
	CreatedAt         time.Time `json:"created_at"`
	ExpiresAt         time.Time `json:"expires_at"`
	CompletenessScore float64   `json:"completeness_score"`
	Checksum          string    `json:"checksum"`
	AccessedCount     int       `json:"accessed_count"`
}

// SnapshotEvidenceEnvelope contains comprehensive evidence for audit trails
type SnapshotEvidenceEnvelope struct {
	SnapshotEvidence   map[string]interface{} `json:"snapshot_evidence"`
	ProcessingEvidence map[string]interface{} `json:"processing_evidence"`
	AuditTrail         SnapshotAuditTrail     `json:"audit_trail"`
}

// SnapshotAuditTrail contains detailed audit information
type SnapshotAuditTrail struct {
	SnapshotCreated time.Time `json:"snapshot_created"`
	WorkflowExecuted time.Time `json:"workflow_executed"`
	DataSources     []string  `json:"data_sources"`
	IntegrityChecks []string  `json:"integrity_checks"`
	ProcessingSteps []string  `json:"processing_steps"`
}

// SnapshotPerformanceMetrics contains performance metrics for snapshot-based execution
type SnapshotPerformanceMetrics struct {
	TotalExecutionTimeMs    int64   `json:"total_execution_time_ms"`
	SnapshotRetrievalTimeMs int64   `json:"snapshot_retrieval_time_ms"`
	ORBEvaluationTimeMs     int64   `json:"orb_evaluation_time_ms"`
	RustExecutionTimeMs     int64   `json:"rust_execution_time_ms"`
	NetworkHops             int     `json:"network_hops"`
	ArchitectureType        string  `json:"architecture_type"`
	DataFreshness          float64 `json:"data_freshness_minutes"`
	IntegrityVerified      bool    `json:"integrity_verified"`
}

// AdvancedSnapshotRequest represents an advanced snapshot workflow request
type AdvancedSnapshotRequest struct {
	PatientID         string   `json:"patient_id" binding:"required"`
	MedicationCode    string   `json:"medication_code" binding:"required"`
	MedicationName    string   `json:"medication_name,omitempty"`
	Indication        string   `json:"indication,omitempty"`
	PatientConditions []string `json:"patient_conditions,omitempty"`
	Priority          string   `json:"priority,omitempty"`
	ProviderID        string   `json:"provider_id,omitempty"`
	EncounterID       string   `json:"encounter_id,omitempty"`
	TTLHours          int      `json:"ttl_hours,omitempty"`
	ForceRefresh      bool     `json:"force_refresh,omitempty"`
}

// BatchSnapshotRequest represents a batch snapshot execution request
type BatchSnapshotRequest struct {
	Requests []SnapshotBasedFlow2Request `json:"requests" binding:"required,min=1,max=10"`
}

// BatchSnapshotResponse represents the response from batch snapshot execution
type BatchSnapshotResponse struct {
	BatchID             string                        `json:"batch_id"`
	TotalRequests       int                           `json:"total_requests"`
	SuccessCount        int                           `json:"success_count"`
	FailureCount        int                           `json:"failure_count"`
	SuccessfulResults   []SnapshotBasedFlow2Response  `json:"successful_results"`
	FailedResults       []BatchFailureResult          `json:"failed_results"`
	StartedAt           time.Time                     `json:"started_at"`
	CompletedAt         time.Time                     `json:"completed_at"`
	TotalExecutionTimeMs int64                        `json:"total_execution_time_ms"`
}

// BatchFailureResult represents a failed request in batch processing
type BatchFailureResult struct {
	Index   int                       `json:"index"`
	Request SnapshotBasedFlow2Request `json:"request"`
	Error   string                    `json:"error"`
}

// SnapshotBasedRustRequest represents a request to Rust engine with snapshot data
type SnapshotBasedRustRequest struct {
	RequestID       string                 `json:"request_id"`
	SnapshotID      string                 `json:"snapshot_id"`
	RecipeID        string                 `json:"recipe_id"`
	MedicationCode  string                 `json:"medication_code"`
	ClinicalData    map[string]interface{} `json:"clinical_data"`
	ProcessingHints map[string]interface{} `json:"processing_hints"`
}

// RustRecipeResponse represents the response from Rust engine recipe execution
type RustRecipeResponse struct {
	RequestID          string              `json:"request_id"`
	MedicationProposal *MedicationProposal `json:"medication_proposal"`
	SafetyStatus       string              `json:"safety_status"`
	ExecutionTimeMs    int64               `json:"execution_time_ms"`
	ProcessingEvidence map[string]interface{} `json:"processing_evidence"`
}

// SignatureMethod constants
const (
	SignatureMethodMock     = "mock"
	SignatureMethodRSA2048  = "rsa-2048"
	SignatureMethodECDSAP256 = "ecdsa-p256"
)

// Snapshot workflow constants
const (
	SnapshotWorkflowTypeBasic    = "basic"
	SnapshotWorkflowTypeAdvanced = "advanced"
	SnapshotWorkflowTypeBatch    = "batch"
)

// SnapshotFilters represents filtering options for listing snapshots
type SnapshotFilters struct {
	PatientID  string `json:"patient_id,omitempty"`
	ProviderID string `json:"provider_id,omitempty"`
	RecipeID   string `json:"recipe_id,omitempty"`
	Status     string `json:"status,omitempty"`
	Limit      int    `json:"limit,omitempty"`
}

// SnapshotSummary represents a summary view of a clinical snapshot
type SnapshotSummary struct {
	ID                string    `json:"id"`
	PatientID         string    `json:"patient_id"`
	RecipeID          string    `json:"recipe_id"`
	Status            string    `json:"status"`
	CreatedAt         time.Time `json:"created_at"`
	ExpiresAt         time.Time `json:"expires_at"`
	CompletenessScore float64   `json:"completeness_score"`
	AccessedCount     int       `json:"accessed_count"`
	ProviderID        string    `json:"provider_id,omitempty"`
	EncounterID       string    `json:"encounter_id,omitempty"`
}

// SnapshotMetrics represents snapshot service performance metrics
type SnapshotMetrics struct {
	TotalSnapshots       int                            `json:"total_snapshots"`
	ActiveSnapshots      int                            `json:"active_snapshots"`
	ExpiredSnapshots     int                            `json:"expired_snapshots"`
	AverageCompleteness  float64                        `json:"average_completeness"`
	AverageTTLHours      float64                        `json:"average_ttl_hours"`
	CreationRatePerHour  float64                        `json:"creation_rate_per_hour"`
	AccessRatePerHour    float64                        `json:"access_rate_per_hour"`
	TopRecipes           []map[string]interface{}       `json:"top_recipes"`
	TopProviders         []map[string]interface{}       `json:"top_providers"`
}

// ServiceStatus represents Context Gateway service status
type ServiceStatus struct {
	Service          string                 `json:"service"`
	Status           string                 `json:"status"`
	Version          string                 `json:"version"`
	Features         []string               `json:"features"`
	Endpoints        []string               `json:"endpoints"`
	CurrentMetrics   map[string]interface{} `json:"current_metrics"`
	Timestamp        string                 `json:"timestamp"`
}

// BatchSnapshotResult represents the result of batch snapshot creation
type BatchSnapshotResult struct {
	TotalRequested int                    `json:"total_requested"`
	Successful     []map[string]interface{} `json:"successful"`
	Failed         []map[string]interface{} `json:"failed"`
	CreatedAt      string                 `json:"created_at"`
}

// CleanupResult represents the result of snapshot cleanup operation
type CleanupResult struct {
	Status       string `json:"status"`
	Message      string `json:"message"`
	DeletedCount int    `json:"deleted_count"`
	CleanedAt    string `json:"cleaned_at"`
}

// Performance improvement constants
const (
	// Target performance improvements with snapshot architecture
	TargetSnapshotRetrievalTimeMs = 5
	TargetORBEvaluationTimeMs     = 1
	TargetRustExecutionTimeMs     = 50
	TargetTotalSnapshotTimeMs     = 100 // vs 250ms traditional
	
	// Data integrity constants
	DefaultTTLHours            = 1
	MaxTTLHours               = 24
	MinCompletenessScore      = 0.7
	RequiredIntegrityChecks   = 2 // checksum + signature
)

File: backend/services/context-service/app/models/snapshot_models.py
"""
Clinical Snapshot Data Models

This module defines the data models for the Recipe Snapshot architecture,
providing immutable clinical snapshots with cryptographic integrity.
"""

from datetime import datetime, timedelta
from typing import Dict, Any, Optional, List
from pydantic import BaseModel, Field
from enum import Enum
import uuid


class SnapshotStatus(str, Enum):
    """Snapshot status enumeration"""
    ACTIVE = "active"
    EXPIRED = "expired"
    INVALIDATED = "invalidated"


class SignatureMethod(str, Enum):
    """Digital signature method enumeration"""
    RSA_2048 = "rsa-2048"
    ECDSA_P256 = "ecdsa-p256"
    MOCK = "mock"  # For development/testing


class SnapshotRequest(BaseModel):
    """Request model for creating a clinical snapshot"""
    patient_id: str = Field(..., description="Patient ID for the snapshot")
    recipe_id: str = Field(..., description="Recipe ID to use for data assembly")
    provider_id: Optional[str] = Field(None, description="Provider ID requesting the snapshot")
    encounter_id: Optional[str] = Field(None, description="Encounter ID associated with the snapshot")
    ttl_hours: int = Field(1, ge=1, le=24, description="Time-to-live in hours (1-24)")
    force_refresh: bool = Field(False, description="Force refresh from data sources")
    signature_method: SignatureMethod = Field(SignatureMethod.MOCK, description="Digital signature method")
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class ClinicalSnapshot(BaseModel):
    """Immutable clinical snapshot with cryptographic integrity"""
    
    # Core identification
    id: str = Field(default_factory=lambda: str(uuid.uuid4()), description="Unique snapshot ID")
    patient_id: str = Field(..., description="Patient ID")
    recipe_id: str = Field(..., description="Recipe ID used for assembly")
    context_id: str = Field(..., description="Original context assembly ID")
    
    # Clinical data
    data: Dict[str, Any] = Field(..., description="Assembled clinical data")
    completeness_score: float = Field(..., ge=0.0, le=1.0, description="Data completeness score")
    
    # Integrity verification
    checksum: str = Field(..., description="SHA-256 checksum of the clinical data")
    signature: str = Field(..., description="Digital signature for authenticity")
    signature_method: SignatureMethod = Field(..., description="Signature method used")
    
    # Lifecycle management
    status: SnapshotStatus = Field(SnapshotStatus.ACTIVE, description="Snapshot status")
    created_at: datetime = Field(default_factory=datetime.utcnow, description="Creation timestamp")
    expires_at: datetime = Field(..., description="Expiration timestamp")
    accessed_count: int = Field(0, description="Number of times accessed")
    last_accessed_at: Optional[datetime] = Field(None, description="Last access timestamp")
    
    # Audit and traceability
    provider_id: Optional[str] = Field(None, description="Provider ID who requested the snapshot")
    encounter_id: Optional[str] = Field(None, description="Associated encounter ID")
    assembly_metadata: Dict[str, Any] = Field(..., description="Original assembly metadata")
    
    # Evidence envelope for clinical safety
    evidence_envelope: Dict[str, Any] = Field(default_factory=dict, description="Clinical evidence and decision trail")
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }
    
    def is_expired(self) -> bool:
        """Check if the snapshot has expired"""
        return datetime.utcnow() > self.expires_at
    
    def is_valid(self) -> bool:
        """Check if the snapshot is valid (not expired or invalidated)"""
        return self.status == SnapshotStatus.ACTIVE and not self.is_expired()
    
    def mark_accessed(self) -> None:
        """Mark the snapshot as accessed"""
        self.accessed_count += 1
        self.last_accessed_at = datetime.utcnow()


class SnapshotValidationResult(BaseModel):
    """Result of snapshot integrity validation"""
    
    snapshot_id: str = Field(..., description="Snapshot ID being validated")
    valid: bool = Field(..., description="Whether the snapshot is valid")
    checksum_valid: bool = Field(..., description="Whether the checksum is valid")
    signature_valid: bool = Field(..., description="Whether the signature is valid")
    not_expired: bool = Field(..., description="Whether the snapshot has not expired")
    
    errors: List[str] = Field(default_factory=list, description="List of validation errors")
    warnings: List[str] = Field(default_factory=list, description="List of validation warnings")
    
    validated_at: datetime = Field(default_factory=datetime.utcnow, description="Validation timestamp")
    validation_duration_ms: float = Field(..., description="Validation duration in milliseconds")
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class SnapshotSummary(BaseModel):
    """Summary information about a clinical snapshot"""
    
    id: str = Field(..., description="Snapshot ID")
    patient_id: str = Field(..., description="Patient ID")
    recipe_id: str = Field(..., description="Recipe ID")
    status: SnapshotStatus = Field(..., description="Snapshot status")
    created_at: datetime = Field(..., description="Creation timestamp")
    expires_at: datetime = Field(..., description="Expiration timestamp")
    completeness_score: float = Field(..., description="Data completeness score")
    accessed_count: int = Field(..., description="Access count")
    provider_id: Optional[str] = Field(None, description="Provider ID")
    encounter_id: Optional[str] = Field(None, description="Encounter ID")
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class SnapshotMetrics(BaseModel):
    """Metrics for snapshot operations"""
    
    total_snapshots: int = Field(..., description="Total number of snapshots")
    active_snapshots: int = Field(..., description="Number of active snapshots")
    expired_snapshots: int = Field(..., description="Number of expired snapshots")
    
    average_completeness: float = Field(..., description="Average completeness score")
    average_ttl_hours: float = Field(..., description="Average TTL in hours")
    
    creation_rate_per_hour: float = Field(..., description="Snapshot creation rate per hour")
    access_rate_per_hour: float = Field(..., description="Snapshot access rate per hour")
    
    top_recipes: List[Dict[str, Any]] = Field(..., description="Most used recipes")
    top_providers: List[Dict[str, Any]] = Field(..., description="Most active providers")
    
    calculated_at: datetime = Field(default_factory=datetime.utcnow, description="Metrics calculation timestamp")
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


# Database collection models for MongoDB
class SnapshotDocument(BaseModel):
    """MongoDB document model for clinical snapshots"""
    
    _id: str = Field(..., alias="id")
    patient_id: str
    recipe_id: str
    context_id: str
    data: Dict[str, Any]
    completeness_score: float
    checksum: str
    signature: str
    signature_method: str
    status: str
    created_at: datetime
    expires_at: datetime
    accessed_count: int
    last_accessed_at: Optional[datetime]
    provider_id: Optional[str]
    encounter_id: Optional[str]
    assembly_metadata: Dict[str, Any]
    evidence_envelope: Dict[str, Any]
    
    # TTL index will be created on expires_at field
    
    class Config:
        allow_population_by_field_name = True
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }